File: telemetry/jetbrains/src/test/kotlin/software/aws/toolkits/telemetry/ParserTest.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.telemetry

import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.Test

class ParserTest {
    @Test
    fun typeMissingDescription() {
        assertThatThrownBy {
            TelemetryParser.parseFiles(
                listOf(),
                """
            {
                "types": [
                    {
                        "name": "result",
                        "allowedValues": ["Succeeded", "Failed", "Cancelled"],
                    }
                ],
                "metrics": []
            }
            """.trimIndent()
            )
        }.hasMessageContaining("required key [description] not found")
    }

    @Test
    fun missingMetricsField() {
        assertThatThrownBy {
            TelemetryParser.parseFiles(listOf(), """{"types": []}""")
        }.hasMessageContaining("required key [metrics] not found")
    }

    @Test
    fun invalidDataTypes() {
        assertThatThrownBy {
            TelemetryParser.parseFiles(
                listOf(),
                """
            {
                "types": [
                    {
                        "name": "result",
                        "type": "type that does not exist",
                        "description": "abc"
                    }
                ],
                "metrics": []
            }
            """.trimIndent()
            )
        }.hasMessageContaining("type that does not exist is not a valid enum value")

    }

    @Test
    fun successfulParse() {
        TelemetryParser.parseFiles(
            listOf(),
            """
            {
                "metrics": []
            }
            """.trimIndent()
        )
    }
}

File: telemetry/vscode/tst/parserTest.ts
/*!
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { validateInput } from '../src/parser'

describe('Parser', () => {
    test('Type missing description fails', () => {
        const input = `{
            "types": [
                {
                    "name": "result",
                    "allowedValues": ["Succeeded", "Failed", "Cancelled"],
                }
            ],
            "metrics": []
        }`
        expect(() => validateInput(input)).toThrowError('Failed to parse')
    })

    test('Missing metrics field', () => {
        const input = `{"types": []}`
        expect(() => validateInput(input)).toThrowError('Failed to parse')
    })

    test('Invalid data types', () => {
        const input = `{
            "types": [
                {
                    "name": "result",
                    "type": "type that does not exist",
                    "description": "abc"
                }
            ],
            "metrics": []
        }`
        expect(() => validateInput(input)).toThrowError('Failed to parse')
    })

    test('Successful parse', () => {
        const input = `{
            "metrics": []
        }`
        validateInput(input)
    })
})
